### ISSUE TITLE
Problem: MySQL aborting transactions under heavy concurrent transfer load

### ISSUE THREAD

**Issue by marktriggs** [artefactual/archivematica#1198]
As a part of investigating rate limiting we have been running tests with lots of concurrent transfers--up to 40 at a time so far.

During one of these test runs, I noticed a transaction was aborted:

```
archivematica-mcp-client_1       | OperationalError: (1213, 'Deadlock found when trying to get lock; try restarting transaction')
```

I'll include the full INNODB status below, but the issue seems to be that the MCP Server is inserting new rows into `Tasks` while the MCP Client is updating `Tasks` to change `startTime`.  The two transactions aren't touching the same rows, but are contending over a lock on the table's primary key.  Mysteriously, MySQL is reporting this as a deadlock, even though it seems like there's no true circular lock holding going on.

Theoretically I think this issue could have happened at any time, but the MCP batching changes probably make it more likely: we're doing a smaller number of longer-running transactions and so there's an increased chance of them overlapping.  Even then, I've only seen the issue once out of a dozen runs, so I thought I'd better fix it while we can see the problem :)

Adding some retry logic to these transactions should do the trick, so I'll send a pull request to introduce that.

INNODB output:

```
------------------------
LATEST DETECTED DEADLOCK
------------------------
2018-07-05 03:02:03 7f466e9c0700
*** (1) TRANSACTION:
TRANSACTION 51228628, ACTIVE 0 sec fetching rows
mysql tables in use 1, locked 1
LOCK WAIT 3 lock struct(s), heap size 360, 3 row lock(s)
MySQL thread id 172, OS thread handle 0x7f466e98f700, query id 2542 172.18.0.12 archivematica updating
UPDATE `Tasks` SET `startTime` = '2018-07-05 03:02:03.642023' WHERE `Tasks`.`taskUUID` IN ('774370e2-11a0-4cc7-8704-8403929b7d56', 'de531f5a-ccea-4b70-8f8e-6f3f24a09d69', '91995a4a-c8c0-4ae1-83c4-5c0395efb12d', 'dfc40425-dbef-4d61-b51d-cd351dd34c22', '64
f4d097-aa5b-453f-b525-6669cbc911a3', 'b82af445-6820-497c-83f8-e9df146e5ae0', 'a9f93eb1-fc16-4d74-95b9-2b67b5444a4c', '8a85e857-c966-4126-b0a1-d9cd6f5184af', '4e575701-b815-4d12-a4b3-9d606d1e2efa', 'bbe1dd20-a936-4544-856b-ce04fe8549df', '08135658-b7fe-40
01-8deb-dd16bae2652f', 'c0908042-21e9-4db0-9de6-12722af63d14', '13ec3edd-11f6-465b-865d-1cf8a6512156', '719a2c19-1eb3-479a-900c-395b9d2cf1c8', 'cb4b0f82-e376-4613-af7c-006dff3e10ce', '97577bb0-fa70-4ea1-83b7-4a0cef977bd1', '90068272-efee-4779-816f-35393e
a43078', 'b85a21d2-9479-4c89-ad20-2f9929c87bf6', '237d50d5-3908-47d0-a7cf-4aa2fa14a2ab', '6345e0eb-8e93-4db3-afe7-b1d30ba578ec', '70b80668-9bfe-4589-865c-b1d3                                                                                               
*** (1) WAITING FOR THIS LOCK TO BE GRANTED:
RECORD LOCKS space id 7676 page no 5 n bits 120 index `PRIMARY` of table `MCP`.`Tasks` trx id 51228628 lock_mode X waiting
*** (2) TRANSACTION:
TRANSACTION 51228574, ACTIVE 1 sec inserting
mysql tables in use 1, locked 1
5 lock struct(s), heap size 1184, 3 row lock(s), undo log entries 128
MySQL thread id 218, OS thread handle 0x7f466e9c0700, query id 2563 172.18.0.10 archivematica update
INSERT INTO `Tasks` (`taskUUID`, `jobuuid`, `createdTime`, `fileUUID`, `fileName`, `exec`, `arguments`, `startTime`, `endTime`, `client`, `stdOut`, `stdError`, `exitCode`) VALUES ('01a2f6ae-4688-4b6f-9d8f-4f232fbedc81', '665c1a89-85e4-4f87-b2aa-bd4e9837e
35c', '2018-07-05 03:02:03.708118', 'None', '%transferDirectory%91776bb2-eb87-441b-818d-ac753c27c419.jpg', 'removeUnneededFiles_v0.0', '\"/var/archivematica/sharedDirectory/currentlyProcessing/test_3/91776bb2-eb87-441b-818d-ac753c27c419.jpg\" \"None\"', 
NULL, NULL, '', '', '', NULL)                                                                                                                                                                                                                                
*** (2) HOLDS THE LOCK(S):
RECORD LOCKS space id 7676 page no 5 n bits 120 index `PRIMARY` of table `MCP`.`Tasks` trx id 51228574 lock_mode X locks rec but not gap
*** (2) WAITING FOR THIS LOCK TO BE GRANTED:
RECORD LOCKS space id 7676 page no 5 n bits 120 index `PRIMARY` of table `MCP`.`Tasks` trx id 51228574 lock_mode X locks gap before rec insert intention waiting
*** WE ROLL BACK TRANSACTION (1)
```

**Comment by nickwilkinson** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Jisc RDSS](https://trello.com/c/onaPPfEg/3-jisc-rdss)

**Comment by jraddaoui** [comment]
Tested with around 50 Transfers at the same time using the Docker Compose environment. On the first try I got several `OperationalError: (1040, 'Too many connections')` exceptions and most of the Tranfers got stuck or failed, without being able to resume them. After increasing the `max_connections` value in the MySQL server from 151 (rare default value) to 1000 I could re-run the same test without issues. 

However, looking at the MCP server and client logs, I could not find any of the [logging messages](https://github.com/artefactual/archivematica/blob/qa/1.x/src/archivematicaCommon/lib/databaseFunctions.py#L372-L377) included in the PR, so it's possible I could not reproduce the initial issue, or maybe I'm looking to the wrong logs.

@sevein, some questions:
- Are those the right logs to look at? Is `archivematicaCommon` logging to somewhere else?
- Do you think we should investigate further the `max_connections` requirement? Or at least document somewhere that it should be increased if you're expecting several simultaneous Transfers?

**Comment by sevein** [comment]
> Are those the right logs to look at? Is archivematicaCommon logging to somewhere else?

Whatever `retryOnFailure` logs should appear in the MCP{Client,Server} logs - I believe you're looking at the right place.

> Do you think we should investigate further the max_connections requirement? Or at least document somewhere that it should be increased if you're expecting several simultaneous Transfers?

I would expect the number of connections created from MCPServer to be somewhat limited, i.e. it should not grow indefinitely like it could have happened in older versions of Archivematica. Every transfer is processed in a separate thread, started as [follows](https://github.com/artefactual/archivematica/blob/8e24b3b4f86b4c1a1f927266c6b7265348b2fc33/src/MCPServer/lib/archivematicaMCP.py#L174):

    Executor.apply_async(createUnitAndJobChain, [path, config], {"terminate": terminate})

`Executor` is a thread pool, the number of threads allocated is determined by `ARCHIVEMATICA_MCPSERVER_PROTOCOL_LIMITTASKTHREADS` which defaults to 75. AFAIK when all threads are busy doing work `apply_async` will block and wait until a thread becomes available again. But then these threads will group batches of tasks and send them to `taskGroupRunner` where another pool of the same size will be making sure that these batches are processed via Gearman. All these threads will be creating their own connections to the database so the numbers can scale quickly! Also you need to add the connections created from MCPClient although I expect the numbers there to be much lower, unless you were running many clients!

So I'm not surprised that you reached the limit but that just reinforces my desire of revisiting the design of MCPServer in Archivematica 1.9. In particular, not having a pool is the first thing we should be tackling. We should be borrowing connections from the pool and return them as soon as we're done, so they can be reused by other threads, instead of holding them forever. Pools are important because connections to the database are very expensive.

Responding to your question, perhaps we should do better at documenting the current limitations like you said. This is still better of what we had in Archivematica 1.7 though! And @joel-simpson is working on documenting the new design in Archivematica 1.8 (see https://github.com/artefactual/archivematica-docs/issues/182). Alternatively (or additionally), we could reduce the default size of the thread pool. 75 threads (times two, because we have two pools) is probably too much considering that we don't have a connection pool yet. The problem with lowering the size too much is that the user wouldn't see a new transfer at all until the pool allocates a new thread, so that's also undesirable. The lack of feedback is usually frustrating.

**Comment by joel-simpson** [comment]
I've done a bit of testing and research on this too. 

First, regarding the original issue: I ran a test with 20 concurrent transfers (each about 75MB with 15 files). I checked the MCP server logs while running and didn't see any issues, and all 20 transfers completed successfully (taking about 35 minutes in total). 
I don't know what is an appropriate volume test for this issue, but Mark originally reported failures on 40 transfers at a time, so I think maybe that would be worth trying 40 next. It is hard to imagine why anyone would be running > 40 concurrent transfers. 

Second, about the `MAX_CONNECTIONS` parameter. The [MySQL documentation](https://dev.mysql.com/doc/refman/8.0/en/client-connections.html#client-connections-volume-management) says: 

> Linux or Solaris should be able to support at least 500 to 1000 simultaneous connections routinely and as many as 10,000 connections if you have many gigabytes of RAM available and the workload from each is low or the response time target undemanding.

That being the case, I can't see why we should not increase the default value to say, 500. I will log that as a new issue, which will also ensure we don't lose the need to document it. In the scalability documentation I described `ARCHIVEMATICA_MCPSERVER_PROTOCOL_LIMITTASKTHREADS` that @sevein mentions above, but there is nothing there right now about `MAX_CONNECTIONS`
